### PR TITLE
Update to consolidate parameter validation logic

### PR DESCRIPTION
**Pull request by m-nash**
Fixes https://github.com/microsoft/typespec/issues/3486

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs [hunk 1]
```diff
@@ -8,8 +8,23 @@
 
 namespace Microsoft.Generator.CSharp.Statements
 {
-    public sealed record DeclareLocalFunctionStatement(CodeWriterDeclaration Name, IReadOnlyList<ParameterProvider> Parameters, CSharpType ReturnType, ValueExpression? BodyExpression, MethodBodyStatement? BodyStatement) : MethodBodyStatement
+    public sealed class DeclareLocalFunctionStatement : MethodBodyStatement
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs:11]: Why was this changed from record?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs [hunk 2]
```diff
@@ -11,38 +13,76 @@ namespace Microsoft.Generator.CSharp.Providers
...
         /// <param name="bodyStatements">The method body.</param>
-        /// <param name="kind">The method kind <see cref="CSharpMethodKinds"/>.</param>
-        public MethodProvider(MethodSignatureBase signature, MethodBodyStatement bodyStatements, CSharpMethodKinds? kind = null)
+        /// <param name="skipParamValidation">Indicates if validation statements for parameters should be added.</param>
+        public MethodProvider(MethodSignatureBase signature, MethodBodyStatement bodyStatements, bool skipParamValidation = false)
         {
             Signature = signature;
-            BodyStatements = bodyStatements;
-            Kind = kind ?? new CSharpMethodKinds("Undefined");
+            skipParamValidation = skipParamValidation || !signature.Modifiers.HasFlag(MethodSignatureModifiers.Public);
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs:30]: Why do we need the skipParamValidation parameter?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Models/Types/MrwSerializationTypeProvider.cs [hunk 3]
```diff
@@ -262,7 +262,8 @@ internal MethodProvider BuildSerializationConstructor()
                 bodyStatements: new MethodBodyStatement[]
                 {
                     GetPropertyInitializers(serializationCtorParameters)
-                });
+                },
+                true);
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Models/Types/MrwSerializationTypeProvider.cs:266]:
```suggestion
                skipParamValidation: true);
```

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs (hunk 2 above)
Comment (JoshLove-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs:30]: Is this for the deserialization constructor? Won't that be internal anyway so it would be handled by the right side of the OR?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/XmlDocProvider.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public XmlDocProvider()
+        {
+            Params = new List<XmlDocParamStatement>();
+            Exceptions = new List<XmlDocExceptionStatement>();
+        }
+
+        public XmlDocSummaryStatement? Summary { get; set; }
+        public IList<XmlDocParamStatement> Params { get; }
+        public XmlDocStatement? Returns { get; }
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/XmlDocProvider.cs:19]: How will this get set?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocExceptionStatement.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        private static string GetText(ParameterValidationType validationType) => validationType switch
+        {
+            ParameterValidationType.AssertNotNull => "is null.",
+            ParameterValidationType.AssertNotNullOrEmpty => "is an empty string, and was expected to be non-empty.",
+            _ => throw new ArgumentOutOfRangeException(nameof(validationType), validationType, $"Cannot create an XmlDocExceptionStatement with {validationType} validationType")
+        };
+
+        internal override void Write(CodeWriter writer)
+        {
+            writer.WritingXmlDocumentation = true;
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocExceptionStatement.cs:41]: I wonder if this could be implemented using something like an XmlWritingScope that implements IDisposable.
```c#

internal class XmlWritingScope : IDisposable 
{
   private readonly CodeWriter _writer;
   public XmlWritingScope(CodeWriter writer)
   {
     _writer = writer;
     _writer .WritingXmlDocumentation = true;
   }

  public void Dispose()
  {
      _writer.WritingXmlDocumentation = false;
  }
```

I guess it would end up being more lines total but at least we abstract away the property.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                if (lineFormat == SingleArgFormat && Lines[0].ArgumentCount == 1 && Lines[0].GetArgument(0) is string strLine)
+                {
+                    stringToCheck = strLine;
+                }
+                string period = stringToCheck.EndsWith(".") ? string.Empty : ".";
+                writer.WriteLine($"/// {StartTag} {Lines[0]}{period} {EndTag}");
+            }
+            else
+            {
+                writer.WriteLine($"/// {StartTag}");
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs:46]: Do we need to auto add the period for the final line?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                }
+                writer.WriteLine($"/// {EndTag}");
+            }
+
+            writer.WritingXmlDocumentation = false;
+        }
+
+        private static bool IsEmpty(IReadOnlyList<FormattableString> lines)
+        {
+            if (lines.Count != 1)
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs:59]: lines.Count of 0 isn't empty?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs (hunk 2 above)
Comment (ArcturusZhang) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs:30]: or maybe we could make this optional parameter to be nullable, and when the value is:
- `true`: we skip the validation
- `false`: we include the validation
- `null`: we apply the default behavior like we are having here: include validation when it is public otherwise no validation

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs [hunk 8]
```diff
@@ -11,38 +13,76 @@ namespace Microsoft.Generator.CSharp.Providers
...
+            {
+                if (parameter.Validation != ParameterValidationType.None)
+                {
+                    yield return Argument.ValidateParameter(parameter);
+                    wroteValidation = true;
+                }
+            }
+            if (wroteValidation)
+                yield return EmptyLineStatement;
+        }
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs:63]: I think maybe we could change this to return a `IReadOnlyList<>` and construct it using a `List` to improve the performance because we are trying to avoid using yield return because of performance issues in other places.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs (hunk 1 above)
Comment (ArcturusZhang) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs:11]: well they do not have to be records because we kind of never need those record features, such as:
- auto generated `ToString` method
- auto generated `Equals` method
one thing we might be using
- auto generated `GetHashCode` method

so changing back to class does not really change anything.
For debugging purpose (because the auto generated ToString might help in debugging purpose), we should try to support it via adding `DebuggerDisplay` attribute on the base type of everything to invoke the write method.

Therefore I am fine with this change.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs [hunk 9]
```diff
@@ -8,8 +8,23 @@
...
+        private DeclareLocalFunctionStatement(CodeWriterDeclaration name, IReadOnlyList<ParameterProvider> parameters, CSharpType returnType, ValueExpression? bodyExpression, MethodBodyStatement? bodyStatement)
+        {
+            Name = name;
+            Parameters = parameters;
+            ReturnType = returnType;
+            BodyExpression = bodyExpression;
+            BodyStatement = bodyStatement;
+        }
+
         internal DeclareLocalFunctionStatement(CodeWriterDeclaration Name, IReadOnlyList<ParameterProvider> Parameters, CSharpType ReturnType, MethodBodyStatement BodyStatement)
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs:28]: so as a naming convention, I think these parameters should not be capitalized

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ForStatement.cs [hunk 10]
```diff
@@ -7,8 +7,19 @@
...
 namespace Microsoft.Generator.CSharp.Statements
 {
-    public sealed record ForStatement(AssignmentExpression? IndexerAssignment, ValueExpression? Condition, ValueExpression? IncrementExpression) : MethodBodyStatement, IEnumerable<MethodBodyStatement>
+    public sealed class ForStatement : MethodBodyStatement, IEnumerable<MethodBodyStatement>
     {
+        public AssignmentExpression? IndexerAssignment { get; set; }
+        public ValueExpression? Condition { get; set; }
+        public ValueExpression? IncrementExpression { get; set; }
+
+        public ForStatement(AssignmentExpression? indexerAssignment, ValueExpression? condition, ValueExpression? incrementExpression)
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ForStatement.cs:16]: i think from the syntax's perspective, this indexerAssignment does not have to be an `AssignmentExpression`?
the for statement is quite flexible and could do a lot of things.
For example, this is possible:
```
for (Console.WriteLine(i); i < 10; i++)
```
it might be meaningless but it is possible.
Considering we are kind of introducing a system of representing csharp syntax, we should not introduce too many constraints.

Comment (ArcturusZhang) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ForStatement.cs:16]: For instance if we have this:
```csharp
	int i = 0;
	for (Console.WriteLine(i); i < 3; i++)
	{
		Console.WriteLine("wow");
	}
```
and it will output:
```
0
wow
wow
wow
```
just as expected.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/IfElseStatement.cs [hunk 11]
```diff
@@ -1,13 +1,21 @@
...
+        public IfStatement If { get; }
+        public MethodBodyStatement? Else { get; }
+
+        public IfElseStatement(IfStatement ifStatement, MethodBodyStatement? elseStatement)
+        {
+            If = ifStatement;
+            Else = elseStatement;
+        }
+
         public IfElseStatement(BoolSnippet condition, MethodBodyStatement ifStatement, MethodBodyStatement? elseStatement, bool inline = false, bool addBraces = true)
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/IfElseStatement.cs:19]: i think we should put `ValueExpression` as the type of `condition` here.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/InvokeStaticMethodStatement.cs [hunk 12]
```diff
@@ -7,13 +7,30 @@
 
 namespace Microsoft.Generator.CSharp.Statements
 {
-    public sealed record InvokeStaticMethodStatement(CSharpType? MethodType, string MethodName, IReadOnlyList<ValueExpression> Arguments, IReadOnlyList<CSharpType>? TypeArguments = null, bool CallAsExtension = false, bool CallAsAsync = false) : MethodBodyStatement
+    public sealed class InvokeStaticMethodStatement : MethodBodyStatement
     {
+        public CSharpType? MethodType { get; }
+        public string MethodName { get; }
+        public IReadOnlyList<ValueExpression> Arguments { get; }
+        public IReadOnlyList<CSharpType>? TypeArguments { get; }
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/InvokeStaticMethodStatement.cs:15]: maybe this does not have to be nullable because empty array could mean exactly that this method invocation does not have type arguments.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/InvokeStaticMethodStatement.cs [hunk 13]
```diff
@@ -7,13 +7,30 @@
...
+        public bool CallAsExtension { get; }
+        public bool CallAsAsync { get; }
+
+        public InvokeStaticMethodStatement(CSharpType? methodType, string methodName, IReadOnlyList<ValueExpression> arguments, IReadOnlyList<CSharpType>? typeArguments = null, bool callAsExtension = false, bool callAsAsync = false)
+        {
+            MethodType = methodType;
+            MethodName = methodName;
+            Arguments = arguments;
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/InvokeStaticMethodStatement.cs:23]: and we could just have
```suggestion
            Arguments = arguments ?? Array.Empty<CSharpType>();
```

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/SingleLineCommentStatement.cs [hunk 14]
```diff
@@ -5,8 +5,15 @@
 
 namespace Microsoft.Generator.CSharp.Statements
 {
-    public sealed record SingleLineCommentStatement(FormattableString Message) : MethodBodyStatement
+    public sealed class SingleLineCommentStatement : MethodBodyStatement
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/SingleLineCommentStatement.cs:8]: well a shortage of this is that
we still cannot put comments in some places, for example our system does not allow we add comments above the `case` in a `switch` like this:
```
switch(something)
{
// case 1
case 1:
break;
}
```

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ThrowStatement.cs [hunk 15]
```diff
@@ -5,8 +5,15 @@
 
 namespace Microsoft.Generator.CSharp.Statements
 {
-    public sealed record ThrowStatement(ValueExpression ThrowExpression) : MethodBodyStatement
+    public sealed class ThrowStatement : MethodBodyStatement
     {
+        public ValueExpression ThrowExpression { get; }
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ThrowStatement.cs:10]: looks like this should not be a **Throw**Expression, it should just be a `expression`. A "throw" expression might indicate this expression has a throw keyword but from the context here, this should just be the thing following keyword `throw`

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ThrowStatement.cs [hunk 16]
```diff
@@ -5,8 +5,15 @@
...
+
+        public ThrowStatement(ValueExpression throwExpression)
+        {
+            ThrowExpression = throwExpression;
+        }
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ThrowStatement.cs:15]: so this is like a wrapper of the `throw expression` to convert it to a statement.

If this is the case, it does not make sense that the type of this expression is just a `ValueExpression`.
What if they do this:
```
var expression = new AssignmentExpression();
var statement = new ThrowStatement(expression); // this will write a thing like `i = 1;` which is not a throw expression at all.
```

I think we should either let it have the same structure as `ThrowExpression`, or let this encapsulate a `ThrowExpression`

Also maybe we need to add an implicit operator from throwExpression as well?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/UnaryOperatorStatement.cs [hunk 17]
```diff
@@ -5,8 +5,15 @@
 
 namespace Microsoft.Generator.CSharp.Statements
 {
-    public sealed record UnaryOperatorStatement(UnaryOperatorExpression Expression) : MethodBodyStatement
+    public sealed class UnaryOperatorStatement : MethodBodyStatement
     {
+        public UnaryOperatorExpression Expression { get; }
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/UnaryOperatorStatement.cs:10]: similar to `ThrowStatement` maybe we need an implicit operator

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs (hunk 2 above)
Comment (JoshLove-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs:30]: As far as I can tell, the only usage is for deserialization constructor which would be internal, but maybe I'm missing a use case.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs (hunk 1 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs:11]: Records are great if you truly just want an immutable object with default everything.  I find them very restrictive when you want to start customizing access on getters / setters so swapped to classes.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs (hunk 2 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs:30]: For us right now yes deserialization ctor is the only time we want to skip or internal methods.  I can drop this param for now until we have a use case outside of that at which point we can always add it back.  Worst case someone can provide their own MethodProvider which doesn't add the validation.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Models/Types/MrwSerializationTypeProvider.cs (hunk 3 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Models/Types/MrwSerializationTypeProvider.cs]: dropped this param so not needed.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs (hunk 6 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs:46]: I think we should leave punctuation in the multiline case up to the producer since we have things like this https://github.com/microsoft/typespec/blob/main/packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/Models/RoundTripModel.cs#L130 which would result in a string period after the last `para` closing tag.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs (hunk 7 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs:63]: renamed method to be more clear what its trying to do

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs [hunk 18]
```diff
@@ -52,7 +45,20 @@ internal override void Write(CodeWriter writer)
             }
         }
 
-        private static bool IsEmpty(IReadOnlyList<FormattableString> lines)
+        private string GetPeriodOrEmpty(FormattableString formattableString)
+        {
+            //should we auto add the '.'?
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/XmlDocStatement.cs:50]: nit: should we change the comment to say we auto add the '.' if not present?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs (hunk 8 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/MethodProvider.cs:62]: Good catch, when I looked at this again I am already creating a hash of parameters that need validation in the xml doc builder so I moved that out so I can use the hash in both places and then I can construct an array of the exact size needed and that seems to be the best perf.

```
|        Method |      Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|----------:|
|   Yieldreturn | 124.90 ns | 2.125 ns | 1.988 ns | 0.0251 |     424 B |
|       UseList | 137.69 ns | 1.360 ns | 1.205 ns | 0.0325 |     544 B |
| UseSingleList |  93.68 ns | 0.748 ns | 0.625 ns | 0.0236 |     400 B |
|      UseArray |  74.51 ns | 1.076 ns | 0.898 ns | 0.0214 |     360 B |
```

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/SingleLineCommentStatement.cs (hunk 14 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/SingleLineCommentStatement.cs:8]: Is this a short coming on this statement all together?  If so since my PR is not adding this can you open / link a follow up issue to track that?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ThrowStatement.cs (hunk 16 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/ThrowStatement.cs:15]: We don't have a ThrowExpression.  Since this is not introduced with my PR can you open / link a follow up item if we think we should tweak this?

**Review by JoshLove-msft — APPROVED** [review]

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs (hunk 1 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/DeclareLocalFunctionStatement.cs:11]: A simple example of this is when you want the property to be non nullable and you want the param to the ctor to be nullable / optional.
